org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/ServiceArgumentProviderTest.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) OSGi Alliance (2020). All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.osgi.test.junit5.service;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceReference;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.test.common.annotation.InjectBundleContext;
+import org.osgi.test.common.dictionary.Dictionaries;
+import org.osgi.test.junit5.context.BundleContextExtension;
+import org.osgi.test.junit5.types.Foo;
+
+@ExtendWith(BundleContextExtension.class)
+public class ServiceArgumentProviderTest {
+
+	@InjectBundleContext
+	static BundleContext			classLevelContext;
+	static ServiceRegistration<Foo>	sr1;
+	static ServiceRegistration<Foo>	sr2;
+	static ServiceRegistration<Foo>	sr3;
+
+	static AtomicInteger			counter	= new AtomicInteger();
+
+	@BeforeAll
+	public static void before(@InjectBundleContext BundleContext bundleContext) {
+
+		sr1 = bundleContext.registerService(Foo.class, new Foo() {}, Dictionaries.dictionaryOf("1", "2"));
+		sr2 = bundleContext.registerService(Foo.class, new Foo() {}, Dictionaries.dictionaryOf("1", "3"));
+		sr3 = bundleContext.registerService(Foo.class, new Foo() {}, Dictionaries.dictionaryOf("1", "4"));
+
+	}
+
```

Comment:
No tests for the other ServiceSource elements?

Reply:
added test on filter and arguments, check execution-count